gateway/modules/crud/mgo/read.go

```diff
@@ -141,3 +183,126 @@ func generateSortOptions(array []string) bson.D {
+					fmt.Sprintf("$%s", function): fmt.Sprintf("$%s", column),
+				}
+			} else {
+				prevGroupStage.(bson.M)[asColumnName] = bson.M{
+					"$sum": 1,
+				}
+			}
+			groupStage = bson.M{
+				"$group": prevGroupStage.(bson.M),
+			}
+			return groupStage
+		}
+		if column != "*" {
+			groupStage = bson.M{
+				"$group": bson.M{
+					"_id": groupByMap,
+					asColumnName: bson.M{
+						fmt.Sprintf("$%s", function): fmt.Sprintf("$%s", column),
+					},
+				},
+			}
+		} else {
+			groupStage = bson.M{
+				"$group": bson.M{
+					"_id": groupByMap,
+					asColumnName: bson.M{
+						"$sum": 1,
+					},
+				},
+			}
+		}
+		return groupStage
+	}
+	return bson.M{"$group": bson.M{"_id": bson.M{}}}
+}
+
+func generateAggregateSortOptions(array []string) bson.M {
+	sort := bson.M{}
+	for _, value := range array {
+		if strings.HasPrefix(value, "-") {
+			sort[strings.TrimPrefix(value, "-")] = -1
+		} else {
+			sort[value] = 1
+		}
+	}
+
+	return sort
+}
+
+func getOptionStage(options *model.ReadOptions) []bson.M {
+	var optionStage []bson.M
+
+	if options.Skip != nil {
+		optionStage = append(optionStage, bson.M{"$skip": options.Skip})
+	}
+	if options.Limit != nil {
+		optionStage = append(optionStage, bson.M{"$limit": options.Limit})
+	}
+	if options.Sort != nil {
+		optionStage = append(optionStage, bson.M{"$sort": generateAggregateSortOptions(options.Sort)})
+	}
+
+	return optionStage
+}
+
+func generateAggregateAsColumnName(function, column string) string {
+	return fmt.Sprintf("%s__%s__%s", utils.GraphQLAggregate, function, column)
+}
+
+func splitAggregateAsColumnName(asColumnName string) (functionName string, columnName string, isAggregateColumn bool) {
+	v := strings.Split(asColumnName, "__")
+	if len(v) != 3 || !strings.HasPrefix(asColumnName, utils.GraphQLAggregate) {
+		return "", "", false
+	}
+	return v[1], v[2], true
+}
+
+func getNestedObject(doc map[string]interface{}) map[string]interface{} {
+	resultObj := make(map[string]map[string]interface{})
+	for asColumnName, value := range doc {
+		functionName, columnName, isAggregateColumn := splitAggregateAsColumnName(asColumnName)
+		if isAggregateColumn {
+			delete(doc, asColumnName)
+			funcValue, ok := resultObj[functionName]
+			if !ok {
+				resultObj[functionName] = map[string]interface{}{columnName: value}
+				continue
+			}
+			funcValue[columnName] = value
+		} else {
+			if _, ok := value.(map[string]interface{}); ok {
+				for key, val := range value.(map[string]interface{}) {
+					delete(doc, asColumnName)
+					doc[key] = val
+				}
+			}
+		}
+	}
+	if len(resultObj) > 0 {
+		doc[utils.GraphQLAggregate] = resultObj
+	}
+	return doc
+}
+
+func generateQuery(pipeline []bson.M, req *model.ReadRequest, asColumnName, function, column string) []bson.M {
+	groupStage := getGroupByStage(pipeline, req.GroupBy, asColumnName, function, column)
+	if len(pipeline) != 2 {
+		pipeline = append(pipeline, groupStage)
+	}
+	return pipeline
+}
```

Comment:
Instead of creating group by stage & function stage in one function you break it down in two function.
  {
    $group: {
       _id: {  key : key, key1 : key1  },
       aggregate__count__age: { $sum: 1 }
    }
  }
This is the typical group stage. break it down in 2 by first only generating the funtion map
{
       aggregate__count__age: { $sum: 1 },
       aggregate__count__score: { $avg 1 }
}
after this stage. another function will generate group stage
  {
    $group: {
       _id: {  key : key, key1 : key1  },
    }
  }
add the first stage in group stage